libdd-profiling-ffi/src/profile_status.rs

```
@@ -0,0 +1,476 @@
+    fn from(value: &'static CStr) -> Self {
+        Self {
+            flags: FLAG_STATIC,
+            err: value.as_ptr(),
+        }
+    }
+}
+
+impl From<CString> for ProfileStatus {
+    fn from(cstring: CString) -> Self {
+        Self {
+            flags: FLAG_ALLOCATED,
+            err: cstring.into_raw(),
+        }
+    }
+}
+
+impl TryFrom<ProfileStatus> for CString {
+    type Error = usize;
+
+    fn try_from(status: ProfileStatus) -> Result<Self, Self::Error> {
+        if status.flags == FLAG_ALLOCATED {
+            Ok(unsafe { CString::from_raw(status.err.cast_mut()) })
+        } else {
+            Err(status.flags)
+        }
+    }
+}
+
+impl TryFrom<&ProfileStatus> for &CStr {
+    type Error = usize;
+
+    fn try_from(status: &ProfileStatus) -> Result<Self, Self::Error> {
+        if status.flags != FLAG_OK {
+            Ok(unsafe { CStr::from_ptr(status.err.cast_mut()) })
+        } else {
+            Err(status.flags)
+        }
+    }
+}
+
+impl From<ProfileStatus> for Result<(), Cow<'static, CStr>> {
+    fn from(status: ProfileStatus) -> Self {
+        let flags = status.flags;
+        let is_error = (flags & MASK_IS_ERROR) != 0;
+        let is_allocated = (flags & MASK_IS_ALLOCATED) != 0;
+        #[allow(clippy::panic)]
+        if cfg!(debug_assertions) && (status.flags & MASK_UNUSED) != 0 {
+            panic!("invalid bit pattern: {flags:b}");
+        }
+        match (is_allocated, is_error) {
+            (false, false) => Ok(()),
+            (false, true) => Err(Cow::Borrowed(unsafe { CStr::from_ptr(status.err) })),
+            (true, true) => Err(Cow::Owned(unsafe {
+                CString::from_raw(status.err.cast_mut())
+            })),
+            (true, false) => {
+                #[allow(clippy::panic)]
+                if cfg!(debug_assertions) {
+                    panic!("invalid bit pattern: {flags:b}");
+                }
+                unsafe { unreachable_unchecked() }
```

What happens if we reach here?

---

Should we instead return an error on a static string?

---

FFI isn't supposed to make these types. In Rust, this combination is not constructed. It doesn't make any sense to have an "allocated" thing which somehow isn't an "error". But you can have an "error" that is not allocated (statically owned). So that's where the asymmetry comes from. I'll add some comments.